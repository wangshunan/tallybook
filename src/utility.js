export const LIST_VIEW = 'list'
export const CHART_VIEW = 'chart'
export const TYPE_INCOME = 'income'
export const TYPE_OUTCOME = 'outcome'
export const TAB_TEXT = [LIST_VIEW, CHART_VIEW]

export const $ = (id) => {
    return document.getElementById(id)
}

export const padLeft = (n) => {
    return n < 10 ? '0' + n : n
}

export const range = (size, startAt = 0) => {
    const arr = []
    for ( let i = 0; i < size; i++ ) {
        arr[i] = startAt + i
    }

    return arr
}

export const flatternArr = (arr) => {
    return arr.reduce((map, item) => {
        map[item.id] = item
        return map
    }, {})
}

export const Colors = {
    blue: '#347eff',
    deepBlue: '#61dafb',
    green: '#28a745',
    red: '#dc3545',
    gray: '#555',
    lightGray: '#efefef',
    white: '#fff'
}

export const ID = () => {
    // Math.random should be unique because of its seeding algorithm.
    // Convert it to base 36 (numbers + letters), and grab the first 9 characters
    // after the decimal.
    return '_' + Math.random().toString(36).substr(2, 9);
}

export const parseToYearAndMonth = (str) => {
    const date = str ? new Date(str) : new Date()
    return {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
    }
}
export const generateChartDataByCategory = (items, type = TYPE_OUTCOME) => {
    let categoryMap = { }
    items.filter(item => item.category.type === type).forEach((item) => {
        if (categoryMap[item.cid]) {
            categoryMap[item.cid].value += (item.price * 1)
            categoryMap[item.cid].items = [...categoryMap[item.cid].items, item.id]
            } else {
            categoryMap[item.cid] = {
                name: item.category.name,
                value: item.price * 1,
                items: [item.id]
            }
        }
    })
    return Object.keys(categoryMap).map(mapKey => ({ ...categoryMap[mapKey] }))
}