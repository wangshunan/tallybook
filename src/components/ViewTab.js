import React from 'react'
import Ionicon from 'react-ionicons'
import PropTypes from 'prop-types'
import {LIST_VIEW, CHART_VIEW} from '../utility'

const generateLinkClass = (current, view) => {
    return current === view ? `nav-link active ${view}` : `nav-link ${view}`
}

const ViewTab = ({activeTab, onTabChange}) => {
    return (
        <ul className="nav nav-tabs nav-fill my-4">
            <li className="nav-item">
                <a                     
                className={generateLinkClass(activeTab, LIST_VIEW)}
                href="#"
                onClick={(e) => {e.preventDefault(); onTabChange(LIST_VIEW)}}
                >
                    <Ionicon
                        className="rounded mr-2"
                        fontSize="25px"
                        color={'#007bff'}
                        icon='ios-paper'
                    />
                    List
                </a>
            </li>
            <li className="nav-item">
                <a 
                className={generateLinkClass(activeTab, CHART_VIEW)}
                href="#"
                onClick={(e) => {e.preventDefault(); onTabChange(CHART_VIEW)}}
                >
                    <Ionicon
                        className="rounded ml-2"
                        fontSize="25px"
                        color={'#007bff'}
                        icon='ios-pie'
                    />
                    Chart
                </a>
            </li>
        </ul>
    )
}

ViewTab.propTypes = {
    activeTab: PropTypes.string.isRequired,
    onTabChange: PropTypes.func.isRequired
}

export default ViewTab