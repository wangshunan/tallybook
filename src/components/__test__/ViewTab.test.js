import React from 'react'
import {LIST_VIEW, CHART_VIEW} from '../../utility'
import { shallow } from 'enzyme'
import ViewTab from '../ViewTab'
import toJson from 'enzyme-to-json';

const props = {
    activeTab: LIST_VIEW,
    onTabChange: jest.fn()
}

describe('test ViewTab component', () => {
    const wrapper = shallow(<ViewTab {...props}/>)

    it('should render the component to match snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('should render correct active tabType', () => {
        expect(wrapper.find(`.${LIST_VIEW}`).hasClass('active')).toBe(true)
    })


    it('should trigger the correct function callbacks', () => {
        const firstItem = wrapper.find('.nav-link').first()
        firstItem.find('a').first().simulate('click', { preventDefault: () => {} })
        expect(props.onTabChange).toHaveBeenCalledWith(LIST_VIEW)

        firstItem.find('a').last().simulate('click', { preventDefault: () => {} })
        expect(props.onTabChange).toHaveBeenCalledWith(LIST_VIEW)
    })
})