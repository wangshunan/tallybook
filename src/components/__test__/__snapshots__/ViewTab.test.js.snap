// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ViewTab component should render the component to match snapshot 1`] = `
<ul
  className="nav nav-tabs nav-fill my-4"
>
  <li
    className="nav-item"
  >
    <a
      className="nav-link active list"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded mr-2"
        color="#007bff"
        fontSize="25px"
        icon="ios-paper"
        rotate={false}
        shake={false}
        style={Object {}}
      />
      List
    </a>
  </li>
  <li
    className="nav-item"
  >
    <a
      className="nav-link chart"
      href="#"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded ml-2"
        color="#007bff"
        fontSize="25px"
        icon="ios-pie"
        rotate={false}
        shake={false}
        style={Object {}}
      />
      Chart
    </a>
  </li>
</ul>
`;
